server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -1662,7 +1629,45 @@ public Volume attachVolumeToVM(Long vmId, Long volumeId, Long deviceId) {
             return vol;
         }
     }
+    
+    public Volume verifyVmAndAttachVolume(Long vmId, VolumeVO volume){
 
+        Account caller = CallContext.current().getCallingAccount();
+        Volume vol = null;
+        if (vmId != null) {
+            // Check that the virtual machine ID is valid and it's a user vm
+            UserVmVO vm = _userVmDao.findById(vmId);
+            if (vm == null || vm.getType() != VirtualMachine.Type.User) {
+                throw new InvalidParameterValueException("Please specify a valid User VM.");
+            }
+
+            // Check that the VM is in the correct state
+            if (vm.getState() != State.Running && vm.getState() != State.Stopped) {
+                throw new InvalidParameterValueException("Please specify a VM that is either running or stopped.");
+            }
+
+            // permission check
```

Comment:
checkPermissions()

Reply:
done...but I think, i'll edit this to make it abit more usable globally

Reply:
done